File: kotlin-archive/jam-pvm/src/main/kotlin/io/forge/jam/pvm/engine/PageMap.kt
package io.forge.jam.pvm.engine

/**
 * - READ_ONLY: Page can be read but not written (R)
 * - READ_WRITE: Page can be read and written (W)
 * - Inaccessible pages are represented by absence in the bitset
 */
enum class PageAccess {
    READ_ONLY,
    READ_WRITE;

    fun isReadable(): Boolean = true
    fun isWritable(): Boolean = this == READ_WRITE
}

/**
 * Per-page access control using efficient bitset operations.
 * Uses two parallel bit arrays (readableBits and writableBits) where each bit
 * represents one memory page. This allows O(1) access checking with minimal
 * memory overhead.
 *
 * @param pageSize The size of each memory page in bytes (typically 4096)
 */
class PageMap(
    val pageSize: UInt
) {
    private val pageSizeShift: Int = pageSize.countTrailingZeroBits()

    // Bitset storage: each Long represents 64 pages
    private var readableBits: LongArray = LongArray(0)
    private var writableBits: LongArray = LongArray(0)
    private var maxPageIndex: UInt = 0u

    companion object {
        private const val BITS_PER_WORD = 64

        /**
         * Creates a PageMap with initial page mappings.
         *
         * @param pageMap List of (address, length, access) tuples defining initial mappings
         * @param pageSize The size of each memory page in bytes
         */
        fun create(
            pageMap: List<Triple<UInt, UInt, PageAccess>>,
            pageSize: UInt
        ): PageMap {
            val map = PageMap(pageSize)

            // First pass: calculate maximum page index needed
            for ((address, length, _) in pageMap) {
                if (length == 0u) continue
                val startIndex = address shr map.pageSizeShift
                val pages = map.numberOfPagesToAccess(address, length.toInt())
                val endIndex = startIndex + pages
                if (endIndex > map.maxPageIndex) {
                    map.maxPageIndex = endIndex
                }
            }

            // Allocate bit arrays
            val wordsNeeded = ((map.maxPageIndex + 63u) / 64u).toInt()
            map.readableBits = LongArray(wordsNeeded)
            map.writableBits = LongArray(wordsNeeded)

            // Second pass: set access bits
            for ((address, length, access) in pageMap) {
                if (length == 0u) continue
                val startIndex = address shr map.pageSizeShift
                val pages = map.numberOfPagesToAccess(address, length.toInt())
                map.setPageAccessRange(startIndex, pages.toInt(), access)
            }

            return map
        }
    }

    /**
     * Calculates the number of pages needed to access a memory range.
     */
    fun numberOfPagesToAccess(address: UInt, length: Int): UInt {
        if (length == 0) return 0u
        val startPage = address shr pageSizeShift
        val endAddress = address + length.toUInt() - 1u
        val endPage = endAddress shr pageSizeShift
        return endPage - startPage + 1u
    }

    /**
     * Ensures the bitset arrays have capacity for the given page index.
     */
    private fun ensureCapacity(pageIndex: UInt) {
        if (pageIndex >= maxPageIndex) {
            maxPageIndex = pageIndex + 1u
            val wordsNeeded = ((maxPageIndex + 63u) / 64u).toInt()
            val currentSize = readableBits.size

            if (wordsNeeded > currentSize) {
                readableBits = readableBits.copyOf(wordsNeeded)
                writableBits = writableBits.copyOf(wordsNeeded)
            }
        }
    }

    /**
     * Checks if a single page is readable.
     */
    private fun isPageReadable(pageIndex: UInt): Boolean {
        val wordIndex = (pageIndex / 64u).toInt()
        if (wordIndex >= readableBits.size) return false
        val bitIndex = (pageIndex % 64u).toInt()
        return (readableBits[wordIndex] and (1L shl bitIndex)) != 0L
    }

    /**
     * Checks if a single page is writable.
     */
    private fun isPageWritable(pageIndex: UInt): Boolean {
        val wordIndex = (pageIndex / 64u).toInt()
        if (wordIndex >= writableBits.size) return false
        val bitIndex = (pageIndex % 64u).toInt()
        return (writableBits[wordIndex] and (1L shl bitIndex)) != 0L
    }

    /**
     * Core function to check if all pages in a range have a certain property.
     * Returns (result, firstFailingPage) where result is true if all pages pass.
     */
    private inline fun checkPagesInRange(
        pageStart: UInt,
        pages: Int,
        bits: LongArray,
        singlePageChecker: (UInt) -> Boolean
    ): Pair<Boolean, UInt> {
        if (pages == 0) return Pair(true, pageStart)

        var currentPage = pageStart
        val pageEnd = pageStart + pages.toUInt()

        while (currentPage < pageEnd) {
            val wordIndex = (currentPage / 64u).toInt()
            val bitIndex = (currentPage % 64u).toInt()
            val bitsInThisWord = minOf(BITS_PER_WORD - bitIndex, (pageEnd - currentPage).toInt())

            // Create mask for the bits in this word
            val mask: Long = if (bitsInThisWord == BITS_PER_WORD) {
                -1L // All bits set
            } else {
                (1L shl bitsInThisWord) - 1L
            }
            val shiftedMask = mask shl bitIndex

            // Check if all required bits are set
            val wordValue = if (wordIndex < bits.size) bits[wordIndex] else 0L
            if ((wordValue and shiftedMask) != shiftedMask) {
                // Some bits aren't set - find which page failed
                for (bit in 0 until bitsInThisWord) {
                    val pageIndex = currentPage + bit.toUInt()
                    if (!singlePageChecker(pageIndex)) {
                        return Pair(false, pageIndex)
                    }
                }
            }

            currentPage += bitsInThisWord.toUInt()
        }

        return Pair(true, pageStart)
    }

    /**
     * Core function to modify bits in a range.
     */
    private inline fun modifyBitsInRange(
        startIndex: UInt,
        pages: UInt,
        modifier: (wordIndex: Int, mask: Long) -> Unit
    ) {
        if (pages == 0u) return

        var currentPage = startIndex
        val endIndex = startIndex + pages

        while (currentPage < endIndex) {
            val wordIndex = (currentPage / 64u).toInt()
            val bitIndex = (currentPage % 64u).toInt()
            val bitsInThisWord = minOf(BITS_PER_WORD - bitIndex, (endIndex - currentPage).toInt())

            val mask: Long = if (bitsInThisWord == BITS_PER_WORD) {
                -1L
            } else {
                (1L shl bitsInThisWord) - 1L
            }
            val shiftedMask = mask shl bitIndex

            modifier(wordIndex, shiftedMask)
            currentPage += bitsInThisWord.toUInt()
        }
    }

    /**
     * Sets access for a range of pages.
     */
    private fun setPageAccessRange(startIndex: UInt, pages: Int, access: PageAccess) {
        if (pages == 0) return

        val endPage = startIndex + pages.toUInt() - 1u
        ensureCapacity(endPage)

        modifyBitsInRange(startIndex, pages.toUInt()) { wordIndex, shiftedMask ->
            when (access) {
                PageAccess.READ_ONLY -> {
                    readableBits[wordIndex] = readableBits[wordIndex] or shiftedMask
                    writableBits[wordIndex] = writableBits[wordIndex] and shiftedMask.inv()
                }

                PageAccess.READ_WRITE -> {
                    readableBits[wordIndex] = readableBits[wordIndex] or shiftedMask
                    writableBits[wordIndex] = writableBits[wordIndex] or shiftedMask
                }
            }
        }
    }

    fun isReadablePages(pageStart: UInt, pages: Int): Pair<Boolean, UInt> {
        return checkPagesInRange(pageStart, pages, readableBits, ::isPageReadable)
    }

    fun isReadable(address: UInt, length: Int): Pair<Boolean, UInt> {
        if (length == 0) return Pair(true, address)
        val startPageIndex = address shr pageSizeShift
        val pages = numberOfPagesToAccess(address, length)
        val (result, page) = isReadablePages(startPageIndex, pages.toInt())
        return Pair(result, page shl pageSizeShift)
    }

    fun isWritablePages(pageStart: UInt, pages: Int): Pair<Boolean, UInt> {
        return checkPagesInRange(pageStart, pages, writableBits, ::isPageWritable)
    }

    fun isWritable(address: UInt, length: Int): Pair<Boolean, UInt> {
        if (length == 0) return Pair(true, address)
        val startPageIndex = address shr pageSizeShift
        val pages = numberOfPagesToAccess(address, length)
        val (result, page) = isWritablePages(startPageIndex, pages.toInt())
        return Pair(result, page shl pageSizeShift)
    }

    fun updatePages(pageIndex: UInt, pages: Int, access: PageAccess) {
        setPageAccessRange(pageIndex, pages, access)
    }

    fun update(address: UInt, length: Int, access: PageAccess) {
        if (length == 0) return
        val startPageIndex = address shr pageSizeShift
        val pages = numberOfPagesToAccess(address, length)
        setPageAccessRange(startPageIndex, pages.toInt(), access)
    }

    fun removeAccessPages(pageIndex: UInt, pages: Int) {
        if (pages == 0) return

        modifyBitsInRange(pageIndex, pages.toUInt()) { wordIndex, shiftedMask ->
            if (wordIndex < readableBits.size) {
                readableBits[wordIndex] = readableBits[wordIndex] and shiftedMask.inv()
                writableBits[wordIndex] = writableBits[wordIndex] and shiftedMask.inv()
            }
        }
    }

    fun removeAccess(address: UInt, length: Int) {
        if (length == 0) return
        val startPageIndex = address shr pageSizeShift
        val pages = numberOfPagesToAccess(address, length)
        removeAccessPages(startPageIndex, pages.toInt())
    }

    fun findGapOrThrow(pages: Int): UInt {
        if (pages == 0) return 0u

        var currentGapStart: UInt? = null
        var gapSize = 0

        // Search a bit beyond maxPageIndex to find gaps after all allocated pages
        val searchLimit = maxOf(maxPageIndex, pages.toUInt() + maxPageIndex + 64u)
        val totalWords = ((searchLimit + 63u) / 64u).toInt()

        for (wordIdx in 0 until totalWords) {
            val readableWord = if (wordIdx < readableBits.size) readableBits[wordIdx] else 0L
            val writableWord = if (wordIdx < writableBits.size) writableBits[wordIdx] else 0L
            val occupiedWord = readableWord or writableWord

            if (occupiedWord == 0L) {
                // Entire word is free
                val pagesInWord = minOf(BITS_PER_WORD, (searchLimit.toInt() - wordIdx * BITS_PER_WORD))
                if (currentGapStart == null) {
                    currentGapStart = (wordIdx * BITS_PER_WORD).toUInt()
                    gapSize = pagesInWord
                } else {
                    gapSize += pagesInWord
                }

                if (gapSize >= pages) {
                    return currentGapStart
                }
                continue
            }

            // Word has some allocated pages - check bit by bit
            val pagesInWord = minOf(BITS_PER_WORD, (searchLimit.toInt() - wordIdx * BITS_PER_WORD))
            for (bitIdx in 0 until pagesInWord) {
                val pageIndex = (wordIdx * BITS_PER_WORD + bitIdx).toUInt()
                val hasAccess = (occupiedWord and (1L shl bitIdx)) != 0L

                if (!hasAccess) {
                    if (currentGapStart == null) {
                        currentGapStart = pageIndex
                        gapSize = 1
                    } else {
                        gapSize++
                    }

                    if (gapSize >= pages) {
                        return currentGapStart
                    }
                } else {
                    currentGapStart = null
                    gapSize = 0
                }
            }
        }

        throw MemoryError.OutOfMemory(0u)
    }

    fun alignToPageStart(address: UInt): UInt {
        return (address shr pageSizeShift) shl pageSizeShift
    }
}

File: modules/pvm/src/main/scala/io/forge/jam/pvm/memory/PageMap.scala
package io.forge.jam.pvm.memory

import spire.math.UInt
import scala.collection.mutable.ArrayBuffer

/**
 * Page access permissions.
 *
 * - ReadOnly: Page can be read but not written (R)
 * - ReadWrite: Page can be read and written (R/W)
 * - Inaccessible pages are represented by absence in the bitset
 */
enum PageAccess:
  case ReadOnly
  case ReadWrite

  def isReadable: Boolean = true
  def isWritable: Boolean = this == ReadWrite

/**
 * Per-page access control using efficient bitset operations.
 *
 * Uses two parallel bit arrays (readableBits and writableBits) where each bit
 * represents one memory page. This allows O(1) access checking with minimal
 * memory overhead.
 *
 * @param pageSize The size of each memory page in bytes (typically 4096)
 */
final class PageMap(val pageSize: UInt):
  private val pageSizeShift: Int = java.lang.Integer.numberOfTrailingZeros(pageSize.signed)

  private val BitsPerWord = 64

  // Bitset storage: each Long represents 64 pages
  private var readableBits: Array[Long] = Array.empty
  private var writableBits: Array[Long] = Array.empty
  private var maxPageIndex: UInt = UInt(0)

  /**
   * Calculates the number of pages needed to access a memory range.
   */
  def numberOfPagesToAccess(address: UInt, length: Int): Int =
    if length == 0 then 0
    else
      val startPage = address.signed >>> pageSizeShift
      val endAddress = address.signed + length - 1
      val endPage = endAddress >>> pageSizeShift
      endPage - startPage + 1

  /**
   * Ensures the bitset arrays have capacity for the given page index.
   */
  private def ensureCapacity(pageIndex: UInt): Unit =
    if pageIndex.signed >= maxPageIndex.signed then
      maxPageIndex = UInt(pageIndex.signed + 1)
      val wordsNeeded = ((maxPageIndex.signed + 63) / 64)
      val currentSize = readableBits.length

      if wordsNeeded > currentSize then
        readableBits = readableBits.padTo(wordsNeeded, 0L)
        writableBits = writableBits.padTo(wordsNeeded, 0L)

  /**
   * Checks if a single page is readable.
   */
  def isPageReadable(pageIndex: UInt): Boolean =
    val wordIndex = pageIndex.signed / 64
    if wordIndex >= readableBits.length then false
    else
      val bitIndex = pageIndex.signed % 64
      (readableBits(wordIndex) & (1L << bitIndex)) != 0L

  /**
   * Checks if a single page is writable.
   */
  def isPageWritable(pageIndex: UInt): Boolean =
    val wordIndex = pageIndex.signed / 64
    if wordIndex >= writableBits.length then false
    else
      val bitIndex = pageIndex.signed % 64
      (writableBits(wordIndex) & (1L << bitIndex)) != 0L

  /**
   * Sets access for a single page.
   */
  def setPageAccess(pageIndex: UInt, access: PageAccess): Unit =
    ensureCapacity(pageIndex)
    val wordIndex = pageIndex.signed / 64
    val bitIndex = pageIndex.signed % 64
    val mask = 1L << bitIndex

    access match
      case PageAccess.ReadOnly =>
        readableBits(wordIndex) = readableBits(wordIndex) | mask
        writableBits(wordIndex) = writableBits(wordIndex) & ~mask
      case PageAccess.ReadWrite =>
        readableBits(wordIndex) = readableBits(wordIndex) | mask
        writableBits(wordIndex) = writableBits(wordIndex) | mask

  /**
   * Removes all access from a single page.
   */
  def removePageAccess(pageIndex: UInt): Unit =
    val wordIndex = pageIndex.signed / 64
    if wordIndex < readableBits.length then
      val bitIndex = pageIndex.signed % 64
      val mask = ~(1L << bitIndex)
      readableBits(wordIndex) = readableBits(wordIndex) & mask
      writableBits(wordIndex) = writableBits(wordIndex) & mask

  /**
   * Sets access for a range of pages.
   */
  def setPageAccessRange(startIndex: UInt, pages: Int, access: PageAccess): Unit =
    if pages <= 0 then return

    val endPage = UInt(startIndex.signed + pages - 1)
    ensureCapacity(endPage)

    modifyBitsInRangeImpl(startIndex, pages) { (wordIndex, shiftedMask) =>
      access match
        case PageAccess.ReadOnly =>
          readableBits(wordIndex) = readableBits(wordIndex) | shiftedMask
          writableBits(wordIndex) = writableBits(wordIndex) & ~shiftedMask
        case PageAccess.ReadWrite =>
          readableBits(wordIndex) = readableBits(wordIndex) | shiftedMask
          writableBits(wordIndex) = writableBits(wordIndex) | shiftedMask
    }

  /**
   * Core function to modify bits in a range.
   */
  private def modifyBitsInRangeImpl(startIndex: UInt, pages: Int)(modifier: (Int, Long) => Unit): Unit =
    if pages <= 0 then return

    var currentPage = startIndex.signed
    val endIndex = startIndex.signed + pages

    while currentPage < endIndex do
      val wordIndex = currentPage / 64
      val bitIndex = currentPage % 64
      val bitsInThisWord = math.min(BitsPerWord - bitIndex, endIndex - currentPage)

      val mask: Long =
        if bitsInThisWord == BitsPerWord then -1L
        else (1L << bitsInThisWord) - 1L
      val shiftedMask = mask << bitIndex

      modifier(wordIndex, shiftedMask)
      currentPage += bitsInThisWord

  /**
   * Checks if all pages in an address range are readable.
   * Returns (isReadable, firstFailingPageAddress).
   */
  def isReadable(address: UInt, length: Int): (Boolean, UInt) =
    if length == 0 then (true, address)
    else
      val startPageIndex = UInt(address.signed >>> pageSizeShift)
      val pages = numberOfPagesToAccess(address, length)
      val (result, failPage) = isReadablePages(startPageIndex, pages)
      (result, UInt(failPage.signed << pageSizeShift))

  /**
   * Checks if all pages in a range are readable.
   */
  def isReadablePages(pageStart: UInt, pages: Int): (Boolean, UInt) =
    checkPagesInRange(pageStart, pages, readableBits, isPageReadable)

  /**
   * Checks if all pages in an address range are writable.
   * Returns (isWritable, firstFailingPageAddress).
   */
  def isWritable(address: UInt, length: Int): (Boolean, UInt) =
    if length == 0 then (true, address)
    else
      val startPageIndex = UInt(address.signed >>> pageSizeShift)
      val pages = numberOfPagesToAccess(address, length)
      val (result, failPage) = isWritablePages(startPageIndex, pages)
      (result, UInt(failPage.signed << pageSizeShift))

  /**
   * Checks if all pages in a range are writable.
   */
  def isWritablePages(pageStart: UInt, pages: Int): (Boolean, UInt) =
    checkPagesInRange(pageStart, pages, writableBits, isPageWritable)

  /**
   * Core function to check if all pages in a range have a certain property.
   * Returns (result, firstFailingPage) where result is true if all pages pass.
   */
  private def checkPagesInRange(
    pageStart: UInt,
    pages: Int,
    bits: Array[Long],
    singlePageChecker: UInt => Boolean
  ): (Boolean, UInt) =
    if pages == 0 then return (true, pageStart)

    var currentPage = pageStart.signed
    val pageEnd = pageStart.signed + pages

    while currentPage < pageEnd do
      val wordIndex = currentPage / 64
      val bitIndex = currentPage % 64
      val bitsInThisWord = math.min(BitsPerWord - bitIndex, pageEnd - currentPage)

      // Create mask for the bits in this word
      val mask: Long =
        if bitsInThisWord == BitsPerWord then -1L
        else (1L << bitsInThisWord) - 1L
      val shiftedMask = mask << bitIndex

      // Check if all required bits are set
      val wordValue = if wordIndex < bits.length then bits(wordIndex) else 0L
      if (wordValue & shiftedMask) != shiftedMask then
        // Some bits aren't set - find which page failed
        var bit = 0
        while bit < bitsInThisWord do
          val pageIndex = UInt(currentPage + bit)
          if !singlePageChecker(pageIndex) then
            return (false, pageIndex)
          bit += 1

      currentPage += bitsInThisWord

    (true, pageStart)

  /**
   * Updates access for a range of pages (by address).
   */
  def update(address: UInt, length: Int, access: PageAccess): Unit =
    if length == 0 then return
    val startPageIndex = UInt(address.signed >>> pageSizeShift)
    val pages = numberOfPagesToAccess(address, length)
    setPageAccessRange(startPageIndex, pages, access)

  /**
   * Updates access for a range of pages (by page index).
   */
  def updatePages(pageIndex: UInt, pages: Int, access: PageAccess): Unit =
    setPageAccessRange(pageIndex, pages, access)

  /**
   * Removes access for a range of pages.
   */
  def removeAccess(address: UInt, length: Int): Unit =
    if length == 0 then return
    val startPageIndex = UInt(address.signed >>> pageSizeShift)
    val pages = numberOfPagesToAccess(address, length)
    removeAccessPages(startPageIndex, pages)

  /**
   * Removes access for a range of pages (by page index).
   */
  def removeAccessPages(pageIndex: UInt, pages: Int): Unit =
    if pages <= 0 then return

    modifyBitsInRangeImpl(pageIndex, pages) { (wordIndex, shiftedMask) =>
      if wordIndex < readableBits.length then
        readableBits(wordIndex) = readableBits(wordIndex) & ~shiftedMask
        writableBits(wordIndex) = writableBits(wordIndex) & ~shiftedMask
    }

  /**
   * Aligns an address to the start of its containing page.
   */
  def alignToPageStart(address: UInt): UInt =
    UInt((address.signed >>> pageSizeShift) << pageSizeShift)

object PageMap:
  /**
   * Creates a PageMap with initial page mappings.
   *
   * @param pageMap List of (address, length, access) tuples defining initial mappings
   * @param pageSize The size of each memory page in bytes
   */
  def create(
    mappings: List[(UInt, UInt, PageAccess)],
    pageSize: UInt
  ): PageMap =
    val map = new PageMap(pageSize)
    val pageSizeShift = java.lang.Integer.numberOfTrailingZeros(pageSize.signed)

    // First pass: calculate maximum page index needed
    for (address, length, _) <- mappings if length.signed > 0 do
      val startIndex = address.signed >>> pageSizeShift
      val pages = map.numberOfPagesToAccess(address, length.signed)
      val endIndex = startIndex + pages
      if endIndex > map.maxPageIndex.signed then
        map.maxPageIndex = UInt(endIndex)

    // Allocate bit arrays
    val wordsNeeded = ((map.maxPageIndex.signed + 63) / 64)
    map.readableBits = new Array[Long](wordsNeeded)
    map.writableBits = new Array[Long](wordsNeeded)

    // Second pass: set access bits
    for (address, length, access) <- mappings if length.signed > 0 do
      val startIndex = UInt(address.signed >>> pageSizeShift)
      val pages = map.numberOfPagesToAccess(address, length.signed)
      map.setPageAccessRange(startIndex, pages, access)

    map
